apps/backend/src/libs/modules/token/base-token.module.ts

```diff
@@ -0,0 +1,48 @@
+import { jwtVerify, SignJWT } from "jose";
+
+import { type JwtPayload } from "~/libs/types/types.js";
+
+type Constructor = {
+	duration: string;
+	encryption: string;
+	secret: Uint8Array;
+};
+
+const SECONDS_TO_MILLISECONDS = 1000;
+const DEFAULT_TIMESTAMP_SECONDS = 0;
+
+class BaseToken {
+	private readonly duration: string;
+	private readonly encryption: string;
+	private readonly secret: Uint8Array;
+
+	constructor({ duration, encryption, secret }: Constructor) {
+		this.duration = duration;
+		this.encryption = encryption;
+		this.secret = secret;
+	}
+
+	async decodeToken(token: string): Promise<JwtPayload> {
+		const { payload } = await jwtVerify(token, this.secret);
+
+		return {
+			expirationTime: new Date(
+				(payload.exp ?? DEFAULT_TIMESTAMP_SECONDS) * SECONDS_TO_MILLISECONDS,
+			).toISOString(),
+			issuedAt: new Date(
+				(payload.iat ?? DEFAULT_TIMESTAMP_SECONDS) * SECONDS_TO_MILLISECONDS,
+			).toISOString(),
+			userId: Number(payload.sub),
```

Comment:
we changed generateToken to accept userId as payload, and removed subject, so I think sub will not work here anymore. We should extract userId like this, or add it to subject as you did initially - up to you 
```suggestion
			userId: Number(payload['userId']),
```